File: Algorithms/Graphs/AStar.js
"use strict";

class Heap { // Insert Actual Heap Here...

    constructor(endPoint) {

        this.items = [];
        this.distSort = (a, b) => dist(a, endPoint) - dist(b, endPoint);

    }

    push(item) {

        this.items.push(item);
        this.items.sort(this.distSort);

    }

    remove() {

        return this.items.shift();

    }

    get size() {

      return this.items.length;

    }

}

Set.prototype.hasPoint = function(point) {

    for (let item of this.values()) {

        if (item.x == point.x && item.y == point.y) {

            return true;

        }

    }

    return false;

};

let inBounds = (grid, p) => Math.min(p.x, p.y) >= 0 && p.x < grid.length && p.y < grid[0].length;

let dist = (a, b) => (a.x - b.x)**2 + (a.y - b.y)**2;

function* getNeighbors(grid, point) {

    let {x, y} = point;

    for (let n of [{x: x + 1, y: y}, {x: x - 1, y: y}, {x: x, y: y + 1}, {x: x, y: y - 1}]) {

        if (inBounds(grid, n) && grid[n.x][n.y] == '.') {

            yield n;

        }

    }

}

let solve = (grid, start, end) => {

    let visited = new Set(),
        parents = new Map(),
        q = new Heap(end);

    q.push(start);

    while (q.size > 0) {

        let cur = q.remove();
        visited.add(cur);

        if (cur.x == end.x && cur.y == end.y) {

            let path = [],
                node = cur;

            while (node != null) {

                path.unshift(node);
                node = parents.get(node);

            }

            return path;

        } else {

            for (let n of getNeighbors(grid, cur)) {

                if (!visited.hasPoint(n)) {

                    q.push(n);
                    parents.set(n, cur);

                }

            }

        }

    }

    return null;

}

let grid = [
    ".####",
    ".#...",
    ".#.#.",
    "...#.",
    ".##.."
];

for (let point of solve(grid, {x: 0, y: 0}, {x: 4, y: 4})) {

    console.log(point);

}

File: Algorithms/Graphs/AStar.py
from heapq import heappush, heappop

def _in_bounds(grid, point):

    r, c = point
    return min(r, c) >= 0 and r < len(grid) and c < len(grid[0])

def _get_neighbors(grid, point):

    r, c = point
    for rr, cc in [[r + 1, c], [r - 1, c], [r, c + 1], [r, c - 1]]:
        if _in_bounds(grid, (rr, cc)) and grid[rr][cc] == '.':
            yield rr, cc

def _dist(a, b):

    return (a[0] - b[0])**2 + (a[1] - b[1])**2

def solve(grid, start, end):

    visited = []
    parents = {}

    q = [(99, start)]

    while len(q) > 0:

        _, cur = heappop(q)
        visited.append(cur)

        if cur == end:

            path = []
            node = cur

            while node != None:
                path.insert(0, node)
                node = parents.get(node)

            return path

        else:

            for n in _get_neighbors(grid, cur):

                if n not in visited:

                    heappush(q, (_dist(n, end), n))
                    parents.update({n : cur})

    return None

def main():

    grid = [
        ".####",
        ".#...",
        ".#.#.",
        "...#.",
        ".##.."
    ]

    for point in solve(grid, (0,0), (4,4)):
        print(point)

if __name__ == '__main__':
    main()
